File: 백준/Gold/16938. 캠프 준비/캠프 준비.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n, L, R, X;
    static int[] array;
    static int answer;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        L = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        X = Integer.parseInt(st.nextToken());

        array = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            array[i] = Integer.parseInt(st.nextToken());
        }

        answer = 0;
        solve(0, 0, Integer.MAX_VALUE, Integer.MIN_VALUE, 0);

        System.out.println(answer);

    }

    private static void solve(int index, int count, int min, int max, int sum) {
        if (index == n) {
            if (2 <= count && L <= sum && sum <= R && X <= max - min) {
                answer++;
            }
            return;
        }

        solve(index + 1, count + 1, Math.min(min, array[index]), Math.max(max, array[index]), sum + array[index]);
        solve(index + 1, count, min, max, sum);
    }
}

File: 백준/Gold/16938. 캠프 준비/캠프 준비.py
import sys

input = sys.stdin.readline

INF = float("inf")


def solve(curr, _min, _max, _sum):
    global answer
    if curr == n:
        if L <= _sum <= R and X <= _max - _min:
            answer += 1
        return

    solve(curr + 1, min(_min, givens[curr]), max(_max, givens[curr]), _sum + givens[curr])
    solve(curr + 1, _min, _max, _sum)


n, L, R, X = map(int, input().split())
givens = list(map(int, input().split()))

answer = 0
solve(0, INF, -INF, 0)

print(answer)